bundles/org.openhab.binding.mpd/src/main/java/org/openhab/binding/mpd/internal/protocol/MPDConnectionThread.java

```diff
@@ -0,0 +1,283 @@
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingStatusDetail;
+import org.openhab.binding.mpd.internal.MPDException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Class for communicating with the music player daemon through a IP connection
+ *
+ * @author Stefan Röllin - Initial contribution
+ */
+@NonNullByDefault
+public class MPDConnectionThread extends Thread {
+
+    private static final int RECONNECTION_TIMEOUT = 60000; // 60 seconds
+
+    private final Logger logger = LoggerFactory.getLogger(MPDConnectionThread.class);
+
+    private final MPDResponseListener listener;
+
+    private final String address;
+    private final Integer port;
+    private final String password;
+
+    private @Nullable Socket socket = null;
+    private @Nullable BufferedReader reader = null;
+    private @Nullable DataOutputStream writer = null;
+
+    private final List<MPDCommand> pendingCommands = new ArrayList<>();
+    private AtomicBoolean isInIdle = new AtomicBoolean(false);
+    private AtomicBoolean disposed = new AtomicBoolean(false);
+
+    public MPDConnectionThread(MPDResponseListener listener, String address, Integer port, String password) {
+        this.listener = listener;
+
+        this.address = address;
+        this.port = port;
+        this.password = password;
+    }
+
+    @Override
+    public void run() {
+
+        try {
+            while (!disposed.get()) {
+                try {
+                    synchronized (pendingCommands) {
+                        pendingCommands.clear();
+                        pendingCommands.add(new MPDCommand("status"));
+                        pendingCommands.add(new MPDCommand("currentsong"));
+                    }
+
+                    establishConnection();
+                    updateThingStatus(ThingStatus.ONLINE, ThingStatusDetail.NONE, null);
+
+                    processPendingCommands();
+                } catch (IOException e) {
+                    updateThingStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+                } catch (MPDException e) {
+                    updateThingStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, e.getMessage());
+                }
+
+                isInIdle.set(false);
+                closeSocket();
+
+                if (!disposed.get()) {
+                    sleep(RECONNECTION_TIMEOUT);
+                }
+            }
+        } catch (InterruptedException e) {
+        }
+    }
+
+    /**
+     * dispose the connection
+     */
+    public void dispose() {
+        disposed.set(true);
+        synchronized (pendingCommands) {
+            pendingCommands.add(new MPDCommand("close"));
```

Comment:
`Thing.dispose()` should return fast. Can't you just close the socket?

Reply:
You are right. I have tried before with closeSocket(), but that did not work, since it blocked somewhere. Now, the socket is closed.